File: etl/calculate_indicators.py
"""
calculate_indicators.py
-----------------------
Hitung indikator teknikal (SMA, RSI, Volume avg) dari tabel ohlcv_harian
menggunakan pandas murni (tanpa pandas-ta / numba), lalu simpan ke
indikator_teknikal_cache.

Jalankan setelah fetch_ohlcv.py selesai.

Cara pakai:
    python calculate_indicators.py [--date 2026-06-24]
"""

import argparse
from datetime import date

import numpy as np
import pandas as pd
from db import get_conn

HORIZON_PARAMS = {
    "harian": {"ma_short": 5,  "ma_long": 10, "rsi_period": 7,  "vol_window": 5},
    "3hari":  {"ma_short": 10, "ma_long": 20, "rsi_period": 9,  "vol_window": 10},
    "5hari":  {"ma_short": 20, "ma_long": 50, "rsi_period": 14, "vol_window": 20},
}


def calc_rsi(series: pd.Series, period: int) -> pd.Series:
    delta = series.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)
    avg_gain = gain.ewm(com=period - 1, min_periods=period).mean()
    avg_loss = loss.ewm(com=period - 1, min_periods=period).mean()
    rs = avg_gain / avg_loss.replace(0, np.nan)
    return 100 - (100 / (1 + rs))


def signal_tren(close, ma_short, ma_long) -> str:
    if any(v is None or (isinstance(v, float) and np.isnan(v)) for v in [close, ma_short, ma_long]):
        return "netral"
    if close > ma_short > ma_long:
        return "baik"
    if close < ma_short and close < ma_long:
        return "tidak_baik"
    return "netral"


def signal_rsi(rsi) -> str:
    if rsi is None or (isinstance(rsi, float) and np.isnan(rsi)):
        return "netral"
    if rsi < 30:
        return "baik"
    if rsi > 70:
        return "tidak_baik"
    if 40 <= rsi <= 60:
        return "baik"
    return "netral"


def signal_volume(volume, vol_avg) -> str:
    if any(v is None or (isinstance(v, float) and np.isnan(v)) for v in [volume, vol_avg]) or vol_avg == 0:
        return "netral"
    return "baik" if volume > vol_avg else "tidak_baik"


def detect_price_action(df: pd.DataFrame) -> str:
    if len(df) < 2:
        return "netral"
    prev = df.iloc[-2]
    curr = df.iloc[-1]
    prev_body = prev["close"] - prev["open"]
    curr_body = curr["close"] - curr["open"]
    curr_range = curr["high"] - curr["low"]

    if (prev_body < 0 and curr_body > 0
            and curr["open"] <= prev["close"]
            and curr["close"] >= prev["open"]):
        return "baik"

    if curr_body > 0 and curr_range > 0 and (curr["open"] - curr["low"]) / curr_range > 0.6:
        return "baik"

    if (prev_body > 0 and curr_body < 0
            and curr["open"] >= prev["close"]
            and curr["close"] <= prev["open"]):
        return "tidak_baik"

    if curr_range > 0 and curr_body < 0 and (curr["high"] - curr["open"]) / curr_range > 0.6:
        return "tidak_baik"

    return "netral"


def get_ohlcv(kode: str, conn) -> pd.DataFrame:
    cur = conn.cursor()
    cur.execute(
        """
        SELECT tanggal, open::float, high::float, low::float, close::float, volume::bigint
        FROM ohlcv_harian
        WHERE kode_saham = %s
        ORDER BY tanggal
        """,
        (kode,),
    )
    rows = cur.fetchall()
    cur.close()
    if not rows:
        return pd.DataFrame()
    df = pd.DataFrame(rows)   # RealDictCursor returns dicts — column names auto-detected
    df["tanggal"] = pd.to_datetime(df["tanggal"])
    return df.set_index("tanggal")


def compute_and_upsert(kode: str, df: pd.DataFrame, trade_date: date, conn):
    if len(df) < 5:
        return
    if df.index[-1].date() != trade_date:
        return

    cur = conn.cursor()
    rows = []
    for horizon, p in HORIZON_PARAMS.items():
        if len(df) < p["ma_long"]:
            continue

        close   = df["close"]
        ma_s    = close.rolling(p["ma_short"]).mean()
        ma_l    = close.rolling(p["ma_long"]).mean()
        rsi     = calc_rsi(close, p["rsi_period"])
        vol_avg = df["volume"].rolling(p["vol_window"]).mean()

        def fval(s):
            v = s.iloc[-1]
            return float(v) if v is not None and not np.isnan(v) else None

        ma_s_val   = fval(ma_s)
        ma_l_val   = fval(ma_l)
        rsi_val    = fval(rsi)
        vol_val    = float(df["volume"].iloc[-1])
        vol_avg_val = fval(vol_avg)
        close_val  = float(close.iloc[-1])

        rows.append((
            kode, trade_date, horizon,
            ma_s_val, ma_l_val, rsi_val, vol_avg_val,
            signal_tren(close_val, ma_s_val, ma_l_val),
            signal_rsi(rsi_val),
            signal_volume(vol_val, vol_avg_val),
            detect_price_action(df.reset_index()),
        ))

    if rows:
        cur.executemany(
            """
            INSERT INTO indikator_teknikal_cache
              (kode_saham, tanggal, horizon, ma_pendek, ma_panjang, rsi, volume_avg_n,
               sinyal_tren, sinyal_momentum, sinyal_volume, sinyal_price_action)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (kode_saham, tanggal, horizon) DO UPDATE SET
                ma_pendek           = EXCLUDED.ma_pendek,
                ma_panjang          = EXCLUDED.ma_panjang,
                rsi                 = EXCLUDED.rsi,
                volume_avg_n        = EXCLUDED.volume_avg_n,
                sinyal_tren         = EXCLUDED.sinyal_tren,
                sinyal_momentum     = EXCLUDED.sinyal_momentum,
                sinyal_volume       = EXCLUDED.sinyal_volume,
                sinyal_price_action = EXCLUDED.sinyal_price_action
            """,
            rows,
        )
        conn.commit()
    cur.close()


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--date", default=str(date.today()))
    args = parser.parse_args()
    trade_date = date.fromisoformat(args.date)

    conn = get_conn()
    cur = conn.cursor()
    cur.execute(
        "SELECT kode_saham FROM emiten WHERE is_syariah = TRUE AND is_bank = FALSE ORDER BY kode_saham"
    )
    universe = [r["kode_saham"] for r in cur.fetchall()]
    cur.close()

    print(f"[calculate_indicators] {len(universe)} emiten, tanggal {trade_date}")
    for i, kode in enumerate(universe):
        df = get_ohlcv(kode, conn)
        if not df.empty:
            compute_and_upsert(kode, df, trade_date, conn)
        if (i + 1) % 50 == 0:
            print(f"  {i+1}/{len(universe)}...")

    conn.close()
    print("[calculate_indicators] Selesai.")


if __name__ == "__main__":
    main()

File: src/lib/etl/calculateIndicators.ts
import { and, eq, inArray, sql } from "drizzle-orm";
import { db } from "@/lib/db/client";
import { emiten, ohlcv_harian, indikator_teknikal_cache } from "@/lib/db/schema";
import { MA_PARAMS } from "@/lib/scoring/weights";
import type { Horizon, Signal } from "@/types";

const HORIZONS: Horizon[] = ["harian", "3hari", "5hari"];

interface Bar {
  tanggal: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

function rollingMean(values: number[], window: number, endIndex: number): number | null {
  const start = endIndex - window + 1;
  if (start < 0) return null;
  let sum = 0;
  for (let i = start; i <= endIndex; i++) sum += values[i];
  return sum / window;
}

// Replikasi pandas .ewm(com=period-1, min_periods=period).mean() (adjust=True default):
// mean_i = N_i / D_i, dengan N_i = x_i + (1-alpha)*N_{i-1}, D_i = 1 + (1-alpha)*D_{i-1}, alpha = 1/period
function calcRsiLast(closes: number[], period: number): number | null {
  const alpha = 1 / period;
  let gNum = 0, gDen = 0, lNum = 0, lDen = 0, count = 0;
  let last: number | null = null;

  for (let i = 1; i < closes.length; i++) {
    const delta = closes[i] - closes[i - 1];
    const gain = Math.max(delta, 0);
    const loss = Math.max(-delta, 0);
    gNum = gain + (1 - alpha) * gNum;
    gDen = 1 + (1 - alpha) * gDen;
    lNum = loss + (1 - alpha) * lNum;
    lDen = 1 + (1 - alpha) * lDen;
    count++;

    if (count >= period) {
      const avgGain = gNum / gDen;
      const avgLoss = lNum / lDen;
      last = avgLoss === 0 ? null : 100 - 100 / (1 + avgGain / avgLoss);
    }
  }
  return last;
}

function signalTren(close: number, maS: number | null, maL: number | null): Signal {
  if (maS === null || maL === null) return "netral";
  if (close > maS && maS > maL) return "baik";
  if (close < maS && close < maL) return "tidak_baik";
  return "netral";
}

function signalRsi(rsi: number | null): Signal {
  if (rsi === null) return "netral";
  if (rsi < 30) return "baik";
  if (rsi > 70) return "tidak_baik";
  if (rsi >= 40 && rsi <= 60) return "baik";
  return "netral";
}

function signalVolume(volume: number, volAvg: number | null): Signal {
  if (volAvg === null || volAvg === 0) return "netral";
  return volume > volAvg ? "baik" : "tidak_baik";
}

function detectPriceAction(bars: Bar[]): Signal {
  if (bars.length < 2) return "netral";
  const prev = bars[bars.length - 2];
  const curr = bars[bars.length - 1];
  const prevBody = prev.close - prev.open;
  const currBody = curr.close - curr.open;
  const currRange = curr.high - curr.low;

  if (prevBody < 0 && currBody > 0 && curr.open <= prev.close && curr.close >= prev.open) return "baik";
  if (currBody > 0 && currRange > 0 && (curr.open - curr.low) / currRange > 0.6) return "baik";
  if (prevBody > 0 && currBody < 0 && curr.open >= prev.close && curr.close <= prev.open) return "tidak_baik";
  if (currRange > 0 && currBody < 0 && (curr.high - curr.open) / currRange > 0.6) return "tidak_baik";
  return "netral";
}

export async function calculateIndicators(tradeDate: string): Promise<{ processed: number }> {
  const universeRows = await db
    .select({ kode_saham: emiten.kode_saham })
    .from(emiten)
    .where(and(eq(emiten.is_syariah, true), eq(emiten.is_bank, false)));
  const kodeList = universeRows.map((r) => r.kode_saham);
  if (kodeList.length === 0) return { processed: 0 };

  const allBars = await db
    .select({
      kode_saham: ohlcv_harian.kode_saham,
      tanggal: ohlcv_harian.tanggal,
      open: sql<number>`${ohlcv_harian.open}::float`,
      high: sql<number>`${ohlcv_harian.high}::float`,
      low: sql<number>`${ohlcv_harian.low}::float`,
      close: sql<number>`${ohlcv_harian.close}::float`,
      volume: sql<number>`${ohlcv_harian.volume}::float`,
    })
    .from(ohlcv_harian)
    .where(inArray(ohlcv_harian.kode_saham, kodeList))
    .orderBy(ohlcv_harian.kode_saham, ohlcv_harian.tanggal);

  const byKode = new Map<string, Bar[]>();
  for (const b of allBars) {
    if (b.close === null) continue;
    const arr = byKode.get(b.kode_saham) ?? [];
    arr.push({ tanggal: b.tanggal, open: b.open, high: b.high, low: b.low, close: b.close, volume: b.volume });
    byKode.set(b.kode_saham, arr);
  }

  type InsertRow = typeof indikator_teknikal_cache.$inferInsert;
  const rows: InsertRow[] = [];
  const processedKodes = new Set<string>();

  for (const kode of kodeList) {
    const bars = byKode.get(kode);
    if (!bars || bars.length < 5) continue;
    if (bars[bars.length - 1].tanggal !== tradeDate) continue;

    const closes = bars.map((b) => b.close);
    const volumes = bars.map((b) => b.volume);
    const lastIdx = bars.length - 1;
    const closeVal = closes[lastIdx];
    const volVal = volumes[lastIdx];
    const priceActionSignal = detectPriceAction(bars);

    for (const horizon of HORIZONS) {
      const p = MA_PARAMS[horizon];
      if (bars.length < p.maPanjang) continue;

      const maS = rollingMean(closes, p.maPendek, lastIdx);
      const maL = rollingMean(closes, p.maPanjang, lastIdx);
      const volAvg = rollingMean(volumes, p.volumeWindow, lastIdx);
      const rsi = calcRsiLast(closes, p.rsiPeriode);

      rows.push({
        kode_saham: kode,
        tanggal: tradeDate,
        horizon,
        ma_pendek: maS !== null ? String(maS) : null,
        ma_panjang: maL !== null ? String(maL) : null,
        rsi: rsi !== null ? String(rsi) : null,
        volume_avg_n: volAvg !== null ? String(volAvg) : null,
        sinyal_tren: signalTren(closeVal, maS, maL),
        sinyal_momentum: signalRsi(rsi),
        sinyal_volume: signalVolume(volVal, volAvg),
        sinyal_price_action: priceActionSignal,
      });
      processedKodes.add(kode);
    }
  }

  const CHUNK = 300;
  for (let i = 0; i < rows.length; i += CHUNK) {
    const chunk = rows.slice(i, i + CHUNK);
    await db
      .insert(indikator_teknikal_cache)
      .values(chunk)
      .onConflictDoUpdate({
        target: [
          indikator_teknikal_cache.kode_saham,
          indikator_teknikal_cache.tanggal,
          indikator_teknikal_cache.horizon,
        ],
        set: {
          ma_pendek: sql`excluded.ma_pendek`,
          ma_panjang: sql`excluded.ma_panjang`,
          rsi: sql`excluded.rsi`,
          volume_avg_n: sql`excluded.volume_avg_n`,
          sinyal_tren: sql`excluded.sinyal_tren`,
          sinyal_momentum: sql`excluded.sinyal_momentum`,
          sinyal_volume: sql`excluded.sinyal_volume`,
          sinyal_price_action: sql`excluded.sinyal_price_action`,
        },
      });
  }

  return { processed: processedKodes.size };
}
